addons/controllers/csi/vspherecsiconfig_utils.go

```diff
@@ -0,0 +1,259 @@
+	if config.Namespace != "" {
+		dvs.VSpherePVCSI.Namespace = config.Namespace
+	}
+
+	if config.ClusterName != "" {
+		dvs.VSpherePVCSI.ClusterName = config.ClusterName
+	}
+
+	if config.ClusterUID != "" {
+		dvs.VSpherePVCSI.ClusterUID = config.ClusterUID
+	}
+
+	if config.SupervisorMasterEndpointHostname != "" {
+		dvs.VSpherePVCSI.SupervisorMasterEndpointHostname = config.SupervisorMasterEndpointHostname
+	}
+
+	if config.SupervisorMasterPort != nil {
+		dvs.VSpherePVCSI.SupervisorMasterPort = *config.SupervisorMasterPort
+	}
+
+	return dvs, nil
+}
+
+func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValuesNonParavirtual(ctx context.Context,
+	vcsiConfig *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (*DataValues, error) {
+	dvs := &DataValues{}
+	config := vcsiConfig.Spec.VSphereCSI.NonParavirtualConfig
+
+	if config == nil {
+		return nil, errors.Errorf("NonParavirtualConfig missing in Spec.VSphereCSI")
+	}
+
+	// populate derived values
+	dvs.VSphereCSI = &DataValuesVSphereCSI{}
+
+	dvs.VSphereCSI.Namespace = tryParseClusterVariableString(ctx, cluster, NamespaceVarName)
+	dvs.VSphereCSI.ClusterName = tryParseClusterVariableString(ctx, cluster, ClusterNameVarName)
+	dvs.VSphereCSI.Server = tryParseClusterVariableString(ctx, cluster, VSphereServerVarName)
+	dvs.VSphereCSI.Datacenter = tryParseClusterVariableString(ctx, cluster, VSphereDatacenterVarName)
+	dvs.VSphereCSI.PublicNetwork = tryParseClusterVariableString(ctx, cluster, VSphereNetworkVarName) //TODO: IS 'VSPHERE_NETWORK' same as 'public network'?
+	dvs.VSphereCSI.Username = tryParseClusterVariableString(ctx, cluster, VSphereUsernameVarName)
+	dvs.VSphereCSI.Password = tryParseClusterVariableString(ctx, cluster, VSpherePasswordVarName)
+	dvs.VSphereCSI.Region = tryParseClusterVariableString(ctx, cluster, VSphereRegionVarName)
+	dvs.VSphereCSI.Zone = tryParseClusterVariableString(ctx, cluster, VSphereZoneVarName)
+	dvs.VSphereCSI.VSphereVersion = tryParseClusterVariableString(ctx, cluster, VSphereVersionVarName)
+	dvs.VSphereCSI.WindowsSupport = tryParseClusterVariableBool(ctx, cluster, IsWindowsWorkloadClusterVarName)
+	dvs.VSphereCSI.TLSThumbprint = tryParseClusterVariableString(ctx, cluster, VSphereTLSThumbprintVarName)
+
+	if cluster.Annotations != nil {
+		dvs.VSphereCSI.HttpProxy = cluster.Annotations[pkgtypes.HTTPProxyConfigAnnotation]
+		dvs.VSphereCSI.HttpsProxy = cluster.Annotations[pkgtypes.HTTPSProxyConfigAnnotation]
+		dvs.VSphereCSI.NoProxy = cluster.Annotations[pkgtypes.NoProxyConfigAnnotation]
+	}
+
+	// populated from default value in https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/vsphere-csi/2.4.1/bundle/config/values.yaml
+	dvs.VSphereCSI.UseTopologyCategories = false
+	dvs.VSphereCSI.ProvisionTimeout = "300s"
+	dvs.VSphereCSI.AttachTimeout = "300s"
+	dvs.VSphereCSI.ResizerTimeout = "300s"
+	dvs.VSphereCSI.DeploymentReplicas = 3
+	dvs.VSphereCSI.InsecureFlag = true
```

Comment:
question:  should we always assign whatever value vsphereCSIConfig ? IIUC, the defaulting thing should be done within `vsphereCSIConfig` (by a webhook who sets all the defaults if not provided)